pkg/controller/serviceexternalip/controller.go

```diff
@@ -171,15 +195,35 @@ func (c *ServiceExternalIPController) Run(stopCh <-chan struct{}) {
 	<-stopCh
 }
 
-// restoreIPAllocations restores the existing external IPs of Services and records the successful ones in ipAllocationMap.
+// restoreIPAllocations restores the existing external IPs of Services and records the successful ones in ipAllocations.
 func (c *ServiceExternalIPController) restoreIPAllocations(services []*corev1.Service) {
 	var previousIPAllocations []externalippool.IPAllocation
+	servicesAllowingSharedIP := sets.New[apimachinerytypes.NamespacedName]()
+	knownIPsByPool := map[string]sets.Set[string]{}
 	for _, svc := range services {
-		ipPool := svc.ObjectMeta.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey]
+		ipPool := getServiceExternalIPPool(svc)
 		if svc.Spec.Type != corev1.ServiceTypeLoadBalancer || ipPool == "" || len(svc.Status.LoadBalancer.Ingress) == 0 {
 			continue
 		}
-		ip := net.ParseIP(svc.Status.LoadBalancer.Ingress[0].IP)
+		ipStr := svc.Status.LoadBalancer.Ingress[0].IP
+		ip := net.ParseIP(ipStr)
+		if ip == nil {
+			continue
+		}
+		if _, exists := knownIPsByPool[ipPool]; !exists {
+			knownIPsByPool[ipPool] = sets.New[string]()
+		}
+		// Another Service already tries to restore the IP allocation, restoring an IP more than once will fail for sure.
+		// Therefore, we just ignore this Service. If both Services allow shared IP, the IP allocation will be restored
+		// by `syncService`.
+		if knownIPsByPool[ipPool].Has(ipStr) {
+			continue
+		}
```

Comment:
Currently it will be released by `syncService`, but we can log unexpected events here.

I'm also considering preserving the IP, just like before restart. Then we can say the annotation affects IP allocation, allocated IP won't be revoked by changing the annotation. What do you think?

Reply:
Based on other discussions, I think my preference would be to preserve IPs just like before restart, and let the user correct the situation. I also don't think it is a good practice for us to mutate the Service and remove a previously-allocated LB IP just because the Agent restarted. Did I understand your question correctly?

Reply:
Yes, we are on the same page.